examples/UNet/massachussets_road_segm.py

```diff
@@ -0,0 +1,340 @@
+
+    prediction_test = lasagne.layers.get_output(output_layer, x_sym, deterministic=True)
+    loss_val = lasagne.objectives.categorical_crossentropy(prediction_test, seg_sym)
+
+    # we multiply our loss by a weight map. In this example the weight map only increases the loss for road pixels and
+    # decreases the loss for other pixels. We do this to ensure that the network puts more focus on getting the roads
+    # right
+    loss_val *= w_sym
+    loss_val = loss_val.mean()
+    loss_val += l2_loss
+    acc = T.mean(T.eq(T.argmax(prediction_test, axis=1), seg_sym), dtype=theano.config.floatX)
+
+    # learning rate has to be a shared variablebecause we decrease it with every epoch
+    params = lasagne.layers.get_all_params(output_layer, trainable=True)
+    learning_rate = theano.shared(np.float32(0.001))
+    updates = lasagne.updates.adam(loss, params, learning_rate=learning_rate)
+
+    train_fn = theano.function([x_sym, seg_sym, w_sym], [loss, acc_train], updates=updates)
+    val_fn = theano.function([x_sym, seg_sym, w_sym], [loss_val, acc])
+    # have a look at plot_some_results if you want  to see how to use pred_fn
+    pred_fn = theano.function([x_sym], prediction_test)
+
+    # this is np.sum(target_train == 0) and np.sum(target_train == 1). No need to compute this every time
+    class_frequencies = np.array([2374093357., 118906643.])
+    # we are taking the log here because we want the net to focus more on the road pixels but not too much (otherwise
+    # it would not be penalized enough for missclassifying terrain pixels which results in too many false positives)
+    class_weights = np.log(class_frequencies[[1,0]])
+    class_weights = class_weights / np.sum(class_weights) * 2.
+    class_weights = class_weights.astype(np.float32)
+
+    # some data augmentation. If you want better results you should invest more effort here. I left rotations and
+    # deformations out for the sake of speed and simplicity
+    train_generator = random_crop_generator(batch_generator(data_train, target_train, BATCH_SIZE), PATCH_SIZE)
+    train_generator = threaded_generator(train_generator, num_cached=10)
+
+    # there is no need for data augmentation on the validation. However we need patches of the same size which is why
+    # we are using the random crop generator here again
+    validation_generator = random_crop_generator(batch_generator(data_valid, target_valid, BATCH_SIZE), PATCH_SIZE)
+    validation_generator = threaded_generator(validation_generator, num_cached=10)
+
+    # do the actual training
+    for epoch in range(N_EPOCHS):
+        print epoch
+        losses_train = []
+        n_batches = 0
+        accuracies_train = []
+        for data, target in train_generator:
+            # the output of the net has shape (BATCH_SIZE, N_CLASSES). We therefore need to flatten the segmentation so
+            # that we can match it with the prediction via the crossentropy loss function
+            target_flat = target.flatten()
```

Comment:
`flatten()` always creates a copy. Prefer `ravel()`. (Again, not a big deal here, but if this is taken to be an educational example...)